File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.cpp
class Solution {
public:
    int maxSumAfterPartitioning(vector<int>& arr, int k) {
        int n = arr.size();
        vector<int> dp(n+1,0);
        for (int i = 0; i < n; i++) {
            int cs = 0;
            int cm = 0;
            for (int j = i; j > max(-1, i - k); j--) {
                if (cm < arr[j]) cm = arr[j];
                int cur = cm * (i - j + 1) + dp[j];
                if (cur > cs) cs = cur;
            }
            dp[i+1] = cs;
        }
        return dp[n];
    }
};

File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.py
class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        n = len(arr)
        dp = [0] * (n + 1)
        for i in range(n):
            cm = cs = 0
            for j in range(i, max(-1, i - k), -1):
                if (cm < arr[j]):
                    cm = arr[j]
                cur = cm * (i - j + 1) + dp[j]
                if cs < cur:
                    cs = cur
            dp[i+1] = cs
        return dp[n]